### PR TITLE
3.0.5 Candidate

### PR DESCRIPTION
**Pull request by bmeike**
This is the Android/Java v3.0.5 release candidate.  Fixes are:
- CBL-3738: Multiple network failures can cause TooManyRequestsException
- CBL-3361: Remove @ExperimentalCoroutines tags from KTX package
- CBL-3716: Array index OOB in postDatabaseChange

Unticketed cleanup:
- Replace expired test certificate
- Remove an extraneous warning in internal Blob constructor
- Correct lock annotations in C4NativePeer

### CODE REVIEW

File: common/main/java/com/couchbase/lite/AbstractDatabase.java [hunk 1]
```diff
@@ -1180,36 +1180,51 @@ private void postDatabaseChanged() {
...
                 }
-                final boolean newExternal = (nChanges > 0) && c4DbChanges[i].isExternal();
 
-                if ((!docIDs.isEmpty()) && ((nChanges <= 0) || (external != newExternal) || (docIDs.size() > 1000))) {
-                    dbChangeNotifier.postChange(new DatabaseChange((Database) this, docIDs));
+                // Core seems to promise that all changes in a batch
+                // will have the same external flag.  :shrug:
+                final boolean newExternal = c4DbChanges.get(0).isExternal();
+
+                if ((docIDs.size() > 1000) || (external != newExternal) ) {
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/AbstractDatabase.java:1199]: (external != newExternal && !docID.isEmpty()) so it doesn't need to do anything if docID is empty.

**Review by pasin — CHANGES_REQUESTED** [review]

File: common/main/java/com/couchbase/lite/AbstractDatabase.java (hunk 1 above)
Comment (borrrden) [reply common/main/java/com/couchbase/lite/AbstractDatabase.java:1199]: Looks like this is handled inside of `postChanges`

**Review by borrrden — CHANGES_REQUESTED** [review]
I think the `start()` issue that was pointed out here is legit.

File: android/main/java/com/couchbase/lite/internal/connectivity/AndroidConnectivityManager.java [hunk 2]
```diff
@@ -70,7 +70,12 @@ public boolean isConnected() {
 
     @Override
     public void registerObserver(@NonNull Observer observer) {
-        synchronized (observers) { observers.put(observer, Boolean.TRUE); }
+         final boolean shouldStart;
+         synchronized (observers) {
+             shouldStart = observers.isEmpty();
+             observers.put(observer, Boolean.TRUE);
+         }
+         if (shouldStart) { start(); }
```
[the remark this answers is not in the document]
Comment (bmeike) [reply android/main/java/com/couchbase/lite/internal/connectivity/AndroidConnectivityManager.java:78]: Yes.  Wow....

File: common/main/java/com/couchbase/lite/AbstractDatabase.java (hunk 1 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/AbstractDatabase.java:1199]: Yeah.... What Jim said.  To me it makes more sense that the thing that handles sending the updates is the thing that check to see if there are any.

**Review by pasin — APPROVED** [review]

**Comment by bmeike** [comment]
Yeh.  The "start" issue is totally legit.  Thanks.

**Pull request merged**